Improved Vietnamese Braille Tables

danghoaiphuc opened the pull request:
Highlights in this update are: cleaned the code, added many more signs, fix translation issues, and added some more test cases.
Updated Vietnamese Braille tables with following notes of each file:
- vi-charsdef.uti: mainly cleaned the code, to to include existing tables.
- vi-lettersdef.uti: removed unnecessary letters, improved/updated and added Greek letter definitions.
- vi-puncsdef.uti: not use en-ueb-math table, and added new signs, moved num indicator definitions to each main tables.
- vi-saigon-g1.ctb: use letter old d and đ for this old table, use its own cap indicator method.
- vi-vn grade 0, 1 and 2: correct cap indicator and improved the tables.
- vi.yaml: updated and added some new test cases.

bertfrees commented on tests/braille-specs/vi.yaml:
```diff
@@ -25,32 +24,23 @@ table:
-  - [3+2(10-5)/6 = 4.16, ⠼⠉⠐⠖⠼⠃⠈⠣⠼⠁⠚⠤⠼⠑⠈⠜⠌⠼⠋ ⠐⠶ ⠼⠙⠲⠁⠋]
+  - [1+2 4-3 5*6 7/8 〉 Ax+B=0, ⠼⠁⠐⠖⠼⠃ ⠼⠙⠤⠼⠉ ⠼⠑⠦⠼⠋ ⠼⠛⠌⠼⠓ ⠐⠕ ⠨⠁⠭⠐⠖⠨⠃⠐⠶⠼⠚]
+  - [3+2(10-5)/6 = 4.16, ⠼⠉⠐⠖⠼⠃⠈⠣⠼⠁⠚⠤⠼⠑⠈⠜⠌⠼⠋ ⠐⠶ ⠼⠙⠄⠁⠋]
+  - [α β γ δ Α Β Γ Δ αβ ΓΔ, ⠰⠁ ⠰⠃ ⠰⠛ ⠰⠙ ⠸⠁ ⠸⠃ ⠸⠛ ⠸⠙ ⠰⠁⠰⠃ ⠸⠛⠸⠙, xfail: "backward translation fails"]
```
It is better to put such tests in a separate section that is run only in forward direction.

bertfrees commented:
Hi @danghoaiphuc. Thank you for the update. It seems you did many improvements to the table However the test file is still quite minimal. Some tables are not even covered at all. Only vi-vn-g0.utb is covered. I would like to see more tests.

egli reviewed (approved):
Looks good, but please add more tests

danghoaiphuc commented:
Hi @bertfrees  and @egli 
I just commited and pushed to my branch but find no way to make pull request for your review. Sorry I am new to Git.
Some updates in my pull request are:
- Added more test cases for 3 Vietnamese tables: vi-vn-g0.utb, vi-saigon-g1.ctb and vi-vn-g1.ctb.
- Added vi-cb8.ctb: this is currently used in Vietnam for users with Braille display. Vietnam hasn't have official computer Braille yet, but we have used this for years. Vietnam computer Braille uses almost all of US Braille chars with single Braille cell, except Vietnamese letters with accents should be defined one on one based on Vietnamese Braille.
I see there is a table named "vi.ctb". I am not sure how it's created but most of Vietnamese chars are not defined correctly. So I propose you should consider to replace with this one. This new table is just to call other two tables: vi-lettersdef.uti and en-us-comp8-ext.utb

danghoaiphuc commented:
Hi @bertfrees
I have updated and commited:
1. vi-cb8.ctb: changed from "both" direction to "forward. If applied one-cell per one Vietnamese letter with accent, it's unable to handle correctly. So Vietnamese user community still prefers to transcribe those letters into two cells. I have brought this to the community for discussion but no official recommendation so far. But the current output of this table is what we expect.
2. vi-puncsdef.uti: already cleaned up unnecessary optcodes with comment
3. vi-saigon-g1.ctb: I know the base and lowercase optcode definitions for letter d, đ and Đ at the beginning of a file logically doesn't need. But it seems a bug of LibLouis. I planned to report but I see that someone has reported recently. That's the issue of not handling the first definition while it has multiple same definitions for a letter. I will update the test cases in the future for remaining tables like computer Braille, this table and the grade 2 full contraction.
All copyright year 2018 of files are updated.
Test file vi.yaml: as we defined all of those characters to output as space. so just removed comment and changed the expected output.

Many thanks for your help to run through the whole check. I already commited and pushed the changes.

bertfrees commented:
@danghoaiphuc Thanks for the prompt response!

> If applied one-cell per one Vietnamese letter with accent, it's unable to handle correctly. So Vietnamese user community still prefers to transcribe those letters into two cells.

OK thanks for confirming. I'm not sure yet what I will do. I will probably add a comment explaining what you said.

> I know the base and lowercase optcode definitions for letter d, đ and Đ at the beginning of a file logically doesn't need. But it seems a bug of LibLouis.

The `lowercase` rules are indeed needed but the `base` rules shouldn't be.

> All copyright year 2018 of files are updated.

2023 should be added, rather than replacing the 2018. All years in which notable changes were done should be in the copyright line, separated with comma's.

bertfrees commented on tables/vi-cb8.ctb:
```diff
@@ -0,0 +1,44 @@
+#-display-name: Vietnamese Computer Braille
+#
+#+language: vi
+#+type: computer
```
I'm gonna add a comment explaining the situation.

bertfrees commented on tables/vi-charsdef.uti:
```diff
@@ -1,9 +1,9 @@
 #-license: LGPLv2.1
 #
-#  Copyright (C) 2018 by Sao Mai Center for the Blind <www.saomaicenter.org/en>
+#  Copyright (C) 2023 by Sao Mai Center for the Blind <www.saomaicenter.org/en>
```
Should say "2018, 2023".

bertfrees commented on tables/vi-saigon-g1.ctb:
```diff
@@ -1,5 +1,5 @@
 #-index-name: Vietnamese, Saigon, contracted
-#-display-name: Southern Vietnamese braille
+#-display-name: Vietnamese Saigon Braille
```
I'm going to put this change in a separate commit. And change the "Braille" to lowercase.

bertfrees commented:
> Test file vi.yaml: as we defined all of those characters to output as space. so just removed comment and changed the expected output.

It doesn't like like you changed something. Will you?

danghoaiphuc commented:
Sorry, that's mine. I wrongly use word "we"...

I just removed base optcode in vi-saigon-g1.ctb. It works as you advice. 
Thanks a lot for this.

And also updated copyright year in files. Again, thanks for your checking.

PS: I am not sure if I need to make another pull request or not, please 
let me know if I need to .

Thanks

On 3/4/2023 5:39 PM, Bert Frees wrote:
>
>     Test file vi.yaml: as we defined all of those characters to output
>     as space. so just removed comment and changed the expected output.
>
> It doesn't like like you changed something. Will you?
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/liblouis/liblouis/pull/1282#issuecomment-1454694642>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AE454LGI6X4JPAUS7MBPGFLW2MLVDANCNFSM6AAAAAAT46LM7E>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

bertfrees commented:
> PS: I am not sure if I need to make another pull request or not, please let me know if I need to

No, it's not needed to make another PR. Everything fine. Thanks!
